core/variable_map.js

```diff
@@ -81,15 +81,21 @@ Blockly.VariableMap.prototype.renameVariable = function(variable, newName) {
   }
 
   if (variableIndex == -1 && newVariableIndex == -1) {
-    this.createVariable(newName, '');
+    var newVar = this.createVariable(newName, '');
     console.log('Tried to rename an non-existent variable.');
   } else if (variableIndex == newVariableIndex ||
       variableIndex != -1 && newVariableIndex == -1) {
     // Only changing case, or renaming to a completely novel name.
-    this.variableMap_[type][variableIndex].name = newName;
+    var renameVar = this.variableMap_[type][variableIndex];
+    Blockly.Events.fire(new Blockly.Events.VarRename(renameVar, newName));
+    renameVar.name = newName;
   } else if (variableIndex != -1 && newVariableIndex != -1) {
     // Renaming one existing variable to another existing variable.
     // The case might have changed, so we update the destination ID.
+    Blockly.Events.fire(new Blockly.Events.VarRename(
+      this.variableMap_[type][newVariableIndex], newName));
```

Comment:
Stash `this.variableMap_[type][newVariableIndex] ` in a variable--you're using it again on line 98, and it's carrying a lot of information to keep inline.